qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/ParameterizedRollingUpgradeTestCase.java

```diff
@@ -0,0 +1,234 @@
+        .setting("xpack.security.enabled", "false")
+        .feature(FeatureFlag.TIME_SERIES_MODE)
+        .build();
+
+    @ClassRule
+    public static TestRule ruleChain = RuleChain.outerRule(repoDirectory).around(cluster);
+
+    @ParametersFactory(shuffle = false)
+    public static Iterable<Object[]> parameters() {
+        final int numNodes = 3;
+        return Stream.concat(Stream.of((Integer) null), IntStream.range(0, numNodes).boxed())
+            .map(n -> new Object[] { n, numNodes })
+            .toList();
+    }
+
+    private static int totalNodes = -1;
+    private static final Set<Integer> upgradedNodes = new HashSet<>();
+    private static boolean upgradeFailed = false;
+    private static IndexVersion oldIndexVersion;
+
+    private final Integer requestedUpgradeNode;
+
+    protected ParameterizedRollingUpgradeTestCase(@Name("upgradeNode") Integer upgradeNode, @Name("totalNodes") int totalNodes) {
+        if (ParameterizedRollingUpgradeTestCase.totalNodes == -1) {
+            ParameterizedRollingUpgradeTestCase.totalNodes = totalNodes;
+        } else {
+            assertThat("The total number of nodes has changed", totalNodes, equalTo(ParameterizedRollingUpgradeTestCase.totalNodes));
+        }
+        this.requestedUpgradeNode = upgradeNode;
+    }
+
+    @Before
+    public void extractOldIndexVersion() throws Exception {
+        if (oldIndexVersion == null && upgradedNodes.isEmpty()) {
+            IndexVersion indexVersion = null;   // these should all be the same version
+
+            Request request = new Request("GET", "_nodes");
+            request.addParameter("filter_path", "nodes.*.index_version,nodes.*.name");
+            Response response = client().performRequest(request);
+            ObjectPath objectPath = ObjectPath.createFromResponse(response);
+            Map<String, Object> nodeMap = objectPath.evaluate("nodes");
+            for (String id : nodeMap.keySet()) {
+                Number ix = objectPath.evaluate("nodes." + id + ".index_version");
+                IndexVersion version;
+                if (ix != null) {
+                    version = IndexVersion.fromId(ix.intValue());
+                } else {
+                    // it doesn't have index version (pre 8.11) - just infer it from the release version
+                    version = IndexVersion.fromId(getOldClusterVersion().id);
+                }
+
+                if (indexVersion == null) {
+                    indexVersion = version;
+                } else {
+                    String name = objectPath.evaluate("nodes." + id + ".name");
+                    assertThat("Node " + name + " has a different index version to other nodes", version, equalTo(indexVersion));
+                }
+            }
+
+            assertThat("Index version could not be read", indexVersion, notNullValue());
+            oldIndexVersion = indexVersion;
+        }
+    }
+
+    @Before
+    public void checkUpgrade() throws Exception {
```

Comment:
nit: maybe have this be named more directly "maybeUpgradeNode", at first I thought this was only verifying the upgrade